src/agent/communicator/src/http_client.cpp

```diff
@@ -0,0 +1,192 @@
+
+        boost::beast::http::request<boost::beast::http::string_body> req {
+            params.method, params.endpoint, HttpVersion1_1};
+        req.set(boost::beast::http::field::host, params.host);
+        req.set(boost::beast::http::field::user_agent, BOOST_BEAST_VERSION_STRING);
+        req.set(boost::beast::http::field::accept, "application/json");
+
+        if (!params.token.empty())
+        {
+            req.set(boost::beast::http::field::authorization, "Bearer " + params.token);
+        }
+
+        if (!params.user_pass.empty())
+        {
+            req.set(boost::beast::http::field::authorization, "Basic " + params.user_pass);
+        }
+
+        if (!params.body.empty())
+        {
+            req.set(boost::beast::http::field::content_type, "application/json");
+            req.body() = params.body;
+            req.prepare_payload();
+        }
+
+        return req;
+    }
+
+    boost::asio::awaitable<void>
+    Co_PerformHttpRequest(boost::asio::ip::tcp::socket& socket,
+                          boost::beast::http::request<boost::beast::http::string_body>& req,
+                          boost::beast::error_code& ec,
+                          std::function<void()> onUnauthorized)
+    {
+        co_await boost::beast::http::async_write(
+            socket, req, boost::asio::redirect_error(boost::asio::use_awaitable, ec));
+        if (ec)
+        {
+            std::cerr << "Error writing request (" << std::to_string(ec.value()) << "): " << ec.message() << std::endl;
+            co_return;
+        }
+
+        boost::beast::flat_buffer buffer;
+        boost::beast::http::response<boost::beast::http::dynamic_body> res;
+        co_await boost::beast::http::async_read(
+            socket, buffer, res, boost::asio::redirect_error(boost::asio::use_awaitable, ec));
+
+        if (ec)
+        {
+            std::cerr << "Error reading response. Response code: " << res.result_int() << std::endl;
+            co_return;
+        }
+
+        if (res.result_int() == static_cast<int>(boost::beast::http::status::unauthorized))
+        {
+            onUnauthorized();
+        }
+
+        std::cout << "Response code: " << res.result_int() << std::endl;
+        std::cout << "Response body: " << boost::beast::buffers_to_string(res.body().data()) << std::endl;
+    }
+
+    boost::asio::awaitable<void> Co_MessageProcessingTask(const std::string& token,
+                                                          HttpRequestParams reqParams,
+                                                          std::function<std::string()> messageGetter,
+                                                          std::function<void()> onUnauthorized)
+    {
+        using namespace std::chrono_literals;
+
+        auto executor = co_await boost::asio::this_coro::executor;
+        boost::asio::steady_timer timer(executor);
+        boost::asio::ip::tcp::resolver resolver(executor);
+
+        while (true)
```

Comment:
Don't we need to check the `m_exitFlag` in this loop like we do it in `WaitForTokenExpirationAndAuthenticate`?

Reply:
I believe that the `m_exitFlag` was a leftover from before this function became a coroutine. When the `io_context` is destroyed the coroutines are not resumed, but the exit status of the application is 0. Maybe we don't need it at all, at least, as it is right now. Of course a more robust termination mechanism will need to be implemented.

Reply:
So if that is the case, I prefer to remove it for now.

Reply:
done 464b21078f7f8a761cd13a2e0c5d0101d2285bf0